metagraph/api/python/metagraph/client.py

```diff
@@ -170,51 +204,135 @@ def ready(self) -> bool:
         for name, graph_client in self.graphs.items():
-            result[name] = graph_client.search(sequence, top_labels,
-                                               discovery_threshold, with_signature,
-                                               align, **align_params)
+            futures[name] = executor.submit(graph_client.search, sequence,
+                                            top_labels, discovery_threshold, with_signature,
+                                            query_coords, count_kmers,
+                                            align, **align_params)
 
-        return result
+        print(f'Made {len(self.graphs)} requests with {num_processes} threads...')
+
+        # Shutdown executor but do not stop futures
+        executor.shutdown(wait=False)
+        return futures
 
     def align(self, sequence: Union[str, Iterable[str]],
+              parallel=True,
               min_exact_match: float = DEFAULT_DISCOVERY_THRESHOLD,
               max_alternative_alignments: int = 1,
               max_num_nodes_per_seq_char: float = DEFAULT_NUM_NODES_PER_SEQ_CHAR) -> Dict[
-        str, pd.DataFrame]:
-        result = {}
+        str, Union[pd.DataFrame, Future]]:
+        """
+        Make align requests with each graph client.
+
+        :param parallel: perform requests in parallel, and instead return Future instances
+        :return: Dict mapping graph names to pd.DataFrame results
+        """
+
+        if not parallel:
+            result = {}
+
+            # Do this iteratively and return once done
+            for name, graph_client in self.graphs.items():
+                result[name] = graph_client.align(sequence, min_exact_match,
+                                                  max_alternative_alignments,
+                                                  max_num_nodes_per_seq_char)
+
+            return result
+
+        # Otherwise, let's query in parallel with a multiprocessing pool
+        num_processes = min(multiprocessing.cpu_count(), len(self.graphs))
+
+        futures = {}
+        executor = ThreadPoolExecutor(max_workers=num_processes)
+
+        # Populate async results dict with concurrent.futures.Future instances
         for name, graph_client in self.graphs.items():
-            # TODO: do this async
-            result[name] = graph_client.align(sequence, min_exact_match,
-                                              max_alternative_alignments,
-                                              max_num_nodes_per_seq_char)
+            futures[name] = executor.submit(graph_client.align, sequence, min_exact_match,
+                                            max_alternative_alignments,
+                                            max_num_nodes_per_seq_char)
 
-        return result
+        print(f'Made {len(self.graphs)} requests with {num_processes} threads...')
+
+        # Shutdown executor but do not stop futures
+        executor.shutdown(wait=False)
+        return futures
```

Comment:
Do we really benefit from returning Futures instead of just always waiting for everything to finish and returning the actual dictionary?